### PR TITLE
Reinstantiate Dynamic Tests and add to CI

### LINKED ISSUE #549: Run Dynamics Tests in CI
Take the tests from https://github.com/nextsimhub/nextsimdg/tree/issue188/dynamics/tests and implement them in the CI pipeline in Github Actions. This will enable to monitor that changes made in the parallelisation (#120, #448, #495 etc) do not change the behaviour of the dynamics code.

### PR DESCRIPTION
**Pull request by MarionBWeinzierl**
# Reinstantiate  Dynamic Tests and add to CI
### Task List
- [x] Retrieve "pacman" tests from https://github.com/nextsimhub/nextsimdg/tree/issue188/dynamics/tests 
- [x] Define the set of test that we want to run and remove the rest
  - [x] Start with example1 and example2
  - [x] Consider whether we need example3 and example4
  - [x] Rename to test and/or something more descriptive
- [x] Make them run again (adapt CMakeLists)
- [x] Add them to the Github Actions workflow file
- [x] Adapt to naming conventions etc.
- [x] Refactor to use doctest
- [x] Updated the README or other documentation

---
# Change Description

This PR closes #549. In order to make sure that the upcoming MPI implementation does not break the dynamics code, and existing implementation of dynamics tests is retrieved from an earlier version of this repository. The CMakeLists file needs adapting to make them run again. Then, they are added to the Continuous Integration implementation in Github Actions. Additional tests already present in the dynamics directory are also added to Github Actions.

The added new/renewed tests test the advection part of the dynamics.

---
# Documentation Impact

The README in the dynamics/test folder has been updated.

---
### Pre-Request Checklist

- [x] The requirements of this pull request are fully captured in an issue or design specification and are linked and summarised in the description of this PR
- [x] No new warnings are generated
- [x] The documentation has been updated (or an issue has been created to track the corresponding change)
- [x] Methods and Tests are commented such that they can be understood without having to obtain additional context
- [x] This PR/Issue is labelled as a bug/feature/enhancement/breaking change
- [ ] File dates have been updated to reflect modification date
- [x] This change conforms to the conventions described in the README

### CODE REVIEW

**Comment by MarionBWeinzierl** [comment]
Info from Thomas:

> I just checked and they are in principle working. And example1 and example 2(don’t know why I called them examples and not tests) should be the ones to start with.
> 
> They both test the advection, 2 is on a domain with periodic boundary data.
> 
> There is a strange automatic check on whether the result is ok and at the moment the tests all fail. But this is due to their   sensitivity. The implementation is correct but there were just slight adjustments over the time that changed the values. Today I’m really busy, but then it’s getter better. I suggest you try to get it to run in your environment and then we can discuss a good way for automatically checking if the results are gut.

**Comment by MarionBWeinzierl** [comment]
@winzerle says the "orcamesh" tests are doing the same as the "sasipmesh" tests (just reading in the mesh), therefore I have removed those.

**Comment by winzerle** [comment]
Hi, 

Both of them are tests of the advection scheme. The second one has periodic boundaries as ’new feature’
So advectiontest is fine and maybe you could add the label periodic to the second one.

I’ll have a look at the readme. I’m sure, it’s mostly old. 

Thomas



* --------------------------------------------
| Prof. Dr. Thomas Richter
| Numerische Mathematik in den Anwendungen
| Otto-von-Guericke Universität Magdeburg
| numerics.ovgu.de
| +49 391 67-57162  +49 176 615 612 26 (mobil)

**Comment by MarionBWeinzierl** [comment]
OK, I seem to have to set the tolerance (the variable wasn't used, btw) to a value > 1 for the tests to pass -- @winzerle , this is probably not what we want? Can you confirm that they all pass for 1e-4 for you? Also, which parameters can I tweak to speed the tests up a bit, maybe cutting some of the tests that do not give additional insight or so?

**Comment by MarionBWeinzierl** [comment]
@TomMelt @timspainNERSC , I assume we want those tests also to adhere to the doctest framework? Or are you happy for them to be added as is atm?

**Comment by MarionBWeinzierl** [comment]
(note that the tests say "passed" atm although the advection ones have failed, as the result is currently just communication over written output)

**Comment by MarionBWeinzierl** [comment]
> @TomMelt @timspainNERSC , I assume we want those tests also to adhere to the doctest framework? Or are you happy for them to be added as is atm?

I decided to implement minimal doctest functionality for the two tests for the time being, and added issue #587 for fully refactoring those tests.

**Comment by MarionBWeinzierl** [comment]
The current status is that all test pass, except from one of the last advection periodic boundary tests on Mac -- it passes fine on Ubuntu. @winzerle is currently trying to look into this, but struggling to get it to build on Mac -- @timspainNERSC could reproduce the error on this Mac, and Thomas has asked for his help now.

**Comment by einola** [comment]
I've run this on my laptop and got the same crash. Attached is a tar file with the VTK output for the test that fails. Does that help @winzerle?

[test2_3_64.tgz](https://github.com/user-attachments/files/16067716/test2_3_64.tgz)

**Comment by winzerle** [comment]
Fixed the tests. 

There was a bug in the limiter. I do not understand why it did something different on Mac and on Linux, but both is wrong. Assume the difference came from some machine precision problem, having a value which is either slightly below or above 1. I corrected the limiter and added new values to the test case. 

For me, it’s now working both on Mac and linux. 

Thomas

* --------------------------------------------
| Prof. Dr. Thomas Richter
| Numerische Mathematik in den Anwendungen
| Otto-von-Guericke Universität Magdeburg
| numerics.ovgu.de
| +49 391 67-57162  +49 176 615 612 26 (mobil)

**Comment by einola** [comment]
It looks like everything's fixed, but the clang-formatter is complaining about lines that don't exist o_o

**Comment by einola** [comment]
I had to merge with develop before I could see the clang-format error. All tests pass, so this should be ready to be merged, right @MarionBWeinzierl ?

**Comment by MarionBWeinzierl** [comment]
> I had to merge with develop before I could see the clang-format error. All tests pass, so this should be ready to be merged, right @MarionBWeinzierl ?

Thank you! Yes, it just requires a review.

**Comment by einola** [comment]
This is very difficult to review because there are so many clang-format changes.

Am I right in assuming that the files to concentrate on are ``Advection_test.cpp`` and ``AdvectionPeriodicBC_test.cpp`` in dynamics/test?

File: dynamics/test/Advection_test.cpp [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+ * @author Thomas Richter <thomas.richter@ovgu.de>
+ */
+
+#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
+#include <doctest/doctest.h>
+
+#include "DGTransport.hpp"
+#include "Interpolations.hpp"
+#include "ParametricMesh.hpp"
+#include "ParametricTools.hpp"
```
Comment (einola) [inline dynamics/test/Advection_test.cpp:13]:
```suggestion
```

ParametricTools.hpp is not used

File: dynamics/test/Advection_test.cpp [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+#include "Tools.hpp"
+#include "dgLimiters.hpp"
+#include "dgVisu.hpp"
+
+#include "../test/testtools.hpp"
+#include "stopwatch.hpp"
```
Comment (einola) [inline dynamics/test/Advection_test.cpp:19]:
```suggestion
```

These two are not used

File: dynamics/test/AdvectionPeriodicBC_test.cpp [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+ * @author Thomas Richter <thomas.richter@ovgu.de>
+ */
+
+ #define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
+ #include <doctest/doctest.h>
+
+#include "DGTransport.hpp"
+#include "Interpolations.hpp"
+#include "ParametricMesh.hpp"
+#include "ParametricTools.hpp"
```
Comment (einola) [inline dynamics/test/AdvectionPeriodicBC_test.cpp:13]:
```suggestion
```

This include is not used

File: dynamics/test/AdvectionPeriodicBC_test.cpp [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+#include "Tools.hpp"
+#include "dgLimit.hpp"
+#include "dgVisu.hpp"
+
+#include "../test/testtools.hpp"
+#include "stopwatch.hpp"
```
Comment (einola) [inline dynamics/test/AdvectionPeriodicBC_test.cpp:19]:
```suggestion
```

These includes are not used

File: dynamics/test/testtools.hpp
Comment (einola) [inline dynamics/test/testtools.hpp:1]: This file is not used

File: dynamics/test/ParametricMesh_test.cpp
Comment (einola) [inline dynamics/test/ParametricMesh_test.cpp:1]: This test is not included in ``CMakeLists.txt``

**Review by einola — COMMENTED** [review]
If @MarionBWeinzierl can confirm that there's no need to look outside dynamics/test, I'm happy to approve this pending minor changes.

File: dynamics/test/ParametricMesh_test.cpp
Comment (MarionBWeinzierl) [reply dynamics/test/ParametricMesh_test.cpp:1]: I added the test, but get a `SIGABRT` error on Ubuntu, and on Mac it doesn't find `Eigen/Dense`. @timspainNERSC , can you help with the Mac error?

File: dynamics/test/testtools.hpp
Comment (MarionBWeinzierl) [reply dynamics/test/testtools.hpp:1]: Removed it

File: dynamics/test/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    "${CoreDir}/${ModelArrayStructure}/ModelArrayDetails.cpp"
+    )
+target_include_directories(testCGModelArray PRIVATE "${CoreDir}" "${SRC_DIR}" "${CoreDir}/${ModelArrayStructure}")
+target_link_libraries(testCGModelArray LINK_PUBLIC doctest::doctest Eigen3::Eigen)
+
+
+add_executable(testAdvection Advection_test.cpp
+ "${CoreDir}/ModelArray.cpp"
+ "${CoreDir}/${ModelArrayStructure}/ModelArrayDetails.cpp"
+ "${SRC_DIR}/stopwatch.cpp"
```
Comment (einola) [inline dynamics/test/CMakeLists.txt:39]:
```suggestion
```

Not needed

File: dynamics/test/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+ "${SRC_DIR}/Interpolations.cpp"
+ "${SRC_DIR}/ParametricMesh.cpp"
```
Comment (einola) [inline dynamics/test/CMakeLists.txt:41]:
```suggestion
```

File: dynamics/test/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+ "${SRC_DIR}/DGTransport.cpp"
+ "${SRC_DIR}/ParametricMap.cpp"
+ "${SRC_DIR}/cgParametricMomentum.cpp"
```
Comment (einola) [inline dynamics/test/CMakeLists.txt:44]:
```suggestion
```

File: dynamics/test/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+ )
+target_include_directories(testAdvection PRIVATE "${CoreDir}" "${SRC_DIR}" "${CoreDir}/${ModelArrayStructure}")
+target_link_libraries(testAdvection LINK_PUBLIC doctest::doctest Eigen3::Eigen)
+
+add_executable(testAdvectionPeriodicBC AdvectionPeriodicBC_test.cpp
+ "${CoreDir}/ModelArray.cpp"
+ "${CoreDir}/${ModelArrayStructure}/ModelArrayDetails.cpp"
+ "${SRC_DIR}/stopwatch.cpp"
```
Comment (einola) [inline dynamics/test/CMakeLists.txt:52]:
```suggestion
```

File: dynamics/test/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+ "${SRC_DIR}/DGTransport.cpp"
+ "${SRC_DIR}/ParametricMap.cpp"
+ "${SRC_DIR}/Interpolations.cpp"
+ "${SRC_DIR}/ParametricMesh.cpp"
+ "${SRC_DIR}/cgParametricMomentum.cpp"
```
Comment (einola) [inline dynamics/test/CMakeLists.txt:57]:
```suggestion
```

**Comment by TomMelt** [comment]
I will look into why the ubuntu test fails with a Core Dump

**Comment by einola** [comment]
I got it to compile and run on my mac. It looks like the test fails because it can't find the file ``25km_NH.smesh``, which is located in ``dynamics/test``. But I don't know how to put this in the right place for the CI to see it. Do you @TomMelt?

**Comment by TomMelt** [comment]
> I got it to compile and run on my mac. It looks like the test fails because it can't find the file `25km_NH.smesh`, which is located in `dynamics/test`. But I don't know how to put this in the right place for the CI to see it. Do you @TomMelt?

I have figured it out. I am in the process of pushing 2 PRs to github. The first will separate the `clang-formatter` changes, and the second fixes the missing smesh file.

It is missing because it wasn't _force added_ i.e., `git add -f`. Because `*.smesh` files are in the `.gitignore` they need to be force added if they are moved/renamed. 

They should be ready to go by end of today.

**Comment by einola** [comment]
> It is missing because it wasn't _force added_ i.e., `git add -f`. Because `*.smesh` files are in the `.gitignore` they need to be force added if they are moved/renamed.

I'm not sure this is the issue. The file is in the repository: https://github.com/nextsimhub/nextsimdg/blob/5d655f7aca4f7ecbf24679a1a011342289b3dd70/dynamics/test/25km_NH.smesh - I thought it was because the test looks for the file in the current working directory, but it may not be there. But you'll figure it out ;)

**Comment by TomMelt** [comment]
> > It is missing because it wasn't _force added_ i.e., `git add -f`. Because `*.smesh` files are in the `.gitignore` they need to be force added if they are moved/renamed.
> 
> I'm not sure this is the issue. The file is in the repository: https://github.com/nextsimhub/nextsimdg/blob/5d655f7aca4f7ecbf24679a1a011342289b3dd70/dynamics/test/25km_NH.smesh - I thought it was because the test looks for the file in the current working directory, but it may not be there. But you'll figure it out ;)

Ah that error was just on my rebased version :sweat_smile: ... Anyway, I also fixed the `CMakeLists.txt` which was missing a copy statement.

https://github.com/nextsimhub/nextsimdg/blob/e3a2d84acfbac871ada17eb0cd44fff561cf362c/dynamics/test/CMakeLists.txt#L25

**Comment by TomMelt** [comment]
@timspainNERSC @einola hopefully this is much nicer to review now. I have rebased off the `clang-format`ted branch  `format-dynamics-code`. 

@timspainNERSC , my plan is not to make any additional changes to #617 , for the main reason I want it to just be changes relating to clang format and the renaming of the tests folder. Happy to make a separate PR to deal with the changes you suggested, but I don't want to add any additional confusion to an already tricky PR :sweat_smile:

**Review by einola — APPROVED** [review]

**Comment by MarionBWeinzierl** [comment]
Thanks, all!

**Pull request merged**